README.md

```diff
@@ -241,13 +241,14 @@ For example, if `WARM_IP_TARGET` is set to 5, then `ipamd` attempts to keep 5 fr
 elastic network interfaces on the node are unable to provide these free addresses, `ipamd` attempts to allocate more interfaces
 until `WARM_IP_TARGET` free IP addresses are available.
 
-**NOTE!** Avoid this setting for large clusters, or if the cluster has high pod churn. Setting it will cause additional calls to the
+**NOTE!** Be careful when setting this for large clusters or clusters with high pod churn. Setting this will cause additional calls to the
 EC2 API and that might cause throttling of the requests. It is strongly suggested to set `MINIMUM_IP_TARGET` when using `WARM_IP_TARGET`.
 
 If both `WARM_IP_TARGET` and `MINIMUM_IP_TARGET` are set, `ipamd` will attempt to meet both constraints.
 This environment variable overrides `WARM_ENI_TARGET` behavior. For a detailed explanation, see
 [`WARM_ENI_TARGET`, `WARM_IP_TARGET` and `MINIMUM_IP_TARGET`](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/docs/eni-and-ip-target.md).
 
+If `WARM_IP_TARGET` is not set and `MINIMUM_IP_TARGET` is set then `ipamd` will be limited to `MINIMUM_IP_TARGET` IPs to assign pods on a given node.
```

Comment:
This is not really the case. If `WARM_IP_TARGET` is `0` or not set, the CNI will use the `WARM_ENI_TARGET` setting instead (that defaults to `1`, meaning a full ENI of secondary IPs). `MINIMUM_IP_TARGET` is just a lower limit when freeing up IPs when `WARM_IP_TARGET` is set.

Reply:
This is not the behavior I saw; what I did was set `MINIMUM_IP_TARGET` to 5 and removed `WARM_IP_TARGET` (also confirmed that `WARM_ENI_TARGET` is not set. With this setup, here is what happened (based on the logs):

- Node starts
- CNI plugin allocates 29 IPs on the primary ENI (maxIPsPerENI = 29)
```
"msg":"Current warm IP stats: target: 0, total: 29, assigned: 0, available: 29, short: 0, over 24"
```
- CNI plugin releases 24 IPs from the primary ENI
```
"Current warm IP stats: target: 0, total: 5, assigned: 0, available: 5, short: 0, over 0"
```
- 6th pod attempted to start and failed and continues to fail until a revert
```
"msg":"AssignPodIPv4Address: Skip ENI eni-07964b901e0e8d402 that does not have available addresses"
```


So from this it seems that the ENI was full, then the IPs were released, then it never attempts to re-allocate IPs even with continued failures to AssignPodIP.

Reply:
I also created https://github.com/aws/amazon-vpc-cni-k8s/issues/1272 as the immediate allocation + release is not necessary.

Reply:
@mogren friendly bump on the logs above; it seems that the behavior I experienced doesn't match what you described.

Reply:
Hi @jacksontj, the allocation is expected. When a new ENI gets added, we always add a full ENI of secondary IPs, since that is just one call to add 1 or the maximum number. The reason is to support scaling up without having to do a call to EC2 for every new pod that gets added for the case where a new pod gets added to the node every 5 seconds or so. When no additional pods gets added to the node, the extra IPs will be detached. The cool down before IPs get freed from a new ENI is currently set at 1 minute IIRC.

Reply:
I definitely understand what you are describing there -- and would agree that having them around for some time would help with the churn. So the issue seems to be around that timeout. From the logs I have (pasted below) it took ~0.5s to release those IPs -- which is so fast that its pointless; but if it stuck around for some time I could definitely see that being valuable. Is this timeout configurable?

```
{"level":"debug","ts":"2020-10-26T21:09:05.319Z","caller":"ipamd/ipamd.go:752","msg":"Current warm IP stats: target: 0, total: 0, assigned: 0, available: 0, short: 5, over 0"}
{"level":"debug","ts":"2020-10-26T21:09:08.541Z","caller":"ipamd/ipamd.go:900","msg":"Current warm IP stats: target: 0, total: 29, assigned: 0, available: 29, short: 0, over 24"}
{"level":"debug","ts":"2020-10-26T21:09:08.541Z","caller":"ipamd/ipamd.go:918","msg":"Current warm IP stats: target: 0, total: 29, assigned: 0, available: 29, short: 0, over 24"}
{"level":"debug","ts":"2020-10-26T21:09:08.541Z","caller":"ipamd/ipamd.go:571","msg":"Current warm IP stats: target: 0, total: 29, assigned: 0, available: 29, short: 0, over 24"}
{"level":"debug","ts":"2020-10-26T21:09:08.541Z","caller":"ipamd/ipamd.go:642","msg":"Current warm IP stats: target: 0, total: 29, assigned: 0, available: 29, short: 0, over 24"}
{"level":"debug","ts":"2020-10-26T21:09:09.068Z","caller":"ipamd/ipamd.go:930","msg":"Current warm IP stats: target: 0, total: 5, assigned: 0, available: 5, short: 0, over 0"}
```

Also to note; this is the first 6 lines of the log after grepping for `Current warm` (the ipamd started immediately preceding this): 
```
Checking for IPAM connectivity ... {"level":"info","ts":"2020-10-26T21:09:01.939Z","caller":"aws-k8s-agent/main.go:30","msg":"Starting L-IPAMD v1.6.3  ..."}
```

Reply:
I see the thumbs up but I'm not sure what the next step here is. The above logs show that it was releasing IPs in <1s which seems significantly faster than expected. Assuming thats faster than expected, should I open another issue for that? Or is there one that covers this?

Regardless of this API call count; I think a documentation change (like the one I'm suggesting in this PR) is still required as the behavior/interaction is unclear/incorrect in the current documentation.

Reply:
Hi @jacksontj 

I tried similar settings as you mentioned, 

IPAMD starts -
```
{"level":"info","ts":"2020-11-05T19:44:12.882Z","caller":"aws-k8s-agent/main.go:28","msg":"Starting L-IPAMD v1.7-ip-dirty  ..."}

{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:301","msg":"Using WARM_ENI_TARGET 0"}
{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:302","msg":"Using WARM_IP_TARGET 0"}
{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:303","msg":"Using MINIMUM_IP_TARGET 5"}
{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:311","msg":"Start node init"}
```
Next IPs are added to datastore
```
{"level":"info","ts":"2020-11-05T19:44:13.524Z","caller":"ipamd/ipamd.go:772","msg":"Added ENI(eni-0d89a3dc91821414b)'s IP 192.168.54.65 to datastore"}
```
As soon as the nodeInit is done IPAMD starts nodeIPPoolManager - https://github.com/aws/amazon-vpc-cni-k8s/blob/88917ec2393c83aa98ccc5121c96fde24f39fd71/cmd/aws-k8s-agent/main.go#L61

Even though initially there is a sleep, it is only for 2.5s 

https://github.com/aws/amazon-vpc-cni-k8s/blob/591c44b6d9974e1f15c39fca6b1f3815cde9bcb0/pkg/ipamd/ipamd.go#L480

```
{"level":"info","ts":"2020-11-05T19:44:13.541Z","caller":"runtime/asm_amd64.s:1373","msg":"Starting NodeIPpoolManager with 2500000000"}
{"level":"info","ts":"2020-11-05T19:44:13.541Z","caller":"runtime/asm_amd64.s:1373","msg":"Started sleeping now with 2500000000"}

{"level":"info","ts":"2020-11-05T19:44:16.041Z","caller":"runtime/asm_amd64.s:1373","msg":"Woke up to call updateIPPoolIfRequired"}
{"level":"info","ts":"2020-11-05T19:44:16.042Z","caller":"runtime/asm_amd64.s:1373","msg":"Started sleeping now with 2500000000"}

{"level":"info","ts":"2020-11-05T19:44:18.542Z","caller":"runtime/asm_amd64.s:1373","msg":"Woke up to call nodeIPPoolReconcile"}
```

The PR https://github.com/aws/amazon-vpc-cni-k8s/pull/1280 which you have opened would definitely help your use case.

Regarding the documentation change, will review and get back asap. Thanks again!

Reply:
Hi @jacksontj 

Regarding this comment - "If `WARM_IP_TARGET` is not set and `MINIMUM_IP_TARGET` is set then `ipamd` will be limited to `MINIMUM_IP_TARGET` IPs to assign pods on a given node." -> I verified this and issue isn't happening. My configs had -

```
{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:301","msg":"Using WARM_ENI_TARGET 0"}
{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:302","msg":"Using WARM_IP_TARGET 0"}
{"level":"debug","ts":"2020-11-05T19:44:13.384Z","caller":"ipamd/ipamd.go:303","msg":"Using MINIMUM_IP_TARGET 5"}
```

Can you please confirm if you had a different config?

Thanks.